rust/azure_iot_operations_connector/src/base_connector/managed_azure_device_registry.rs

```diff
@@ -0,0 +1,242 @@
+    /// Receives a notification for a newly created asset. This notification includes
+    /// the [`ManagedAsset`], a [`ManagedAssetUpdateObservation`] to observe for updates on
+    /// the new Asset, and a [`AssetDeletionToken`] to observe for deletion of this Asset
+    #[allow(clippy::unused_async)]
+    pub async fn recv_notification(
+        &self,
+    ) -> Option<(
+        ManagedAsset,
+        ManagedAssetUpdateObservation,
+        AssetDeletionToken,
+    )> {
+        // handle the notification
+        // add asset update observation
+        // create copy of asset with asset and dataset connector functionality (status reporting, data forwarding, create data transformers)
+        None
+    }
+}
+
+/// An Observation for asset update events that uses
+/// multiple underlying clients to get full information for a
+/// [`ProtocolTranslator`] to use.
+pub struct ManagedAssetUpdateObservation {
+    _asset_update_observation: AssetUpdateObservation,
+}
+impl ManagedAssetUpdateObservation {
+    /// Receives an updated [`ManagedAsset`] or [`None`] if there will be no more notifications.
+    ///
+    /// If there are notifications:
+    /// - Returns Some([`ManagedAsset`], [`Option<AckToken>`]) on success
+    ///     - If auto ack is disabled, the [`AckToken`] should be used or dropped when you want the ack to occur. If auto ack is enabled, you may use ([`ManagedAsset`], _) to ignore the [`AckToken`].
+    ///
+    /// A received notification can be acknowledged via the [`AckToken`] by calling [`AckToken::ack`] or dropping the [`AckToken`].
+    #[allow(clippy::unused_async)]
+    pub async fn recv_notification(&self) -> Option<(ManagedAsset, Option<AckToken>)> {
+        // handle the notification
+        None
+    }
+}
+
+/// Azure Device Registry Asset that includes additional functionality
+/// to report status, translate data, and send data to the destination
+pub struct ManagedAsset {
+    // re-export of adr::Asset, but Dataset/Event/etc structs are of type ConnectorDataset/etc
+    /// Asset Definition
+    pub asset_definition: Asset,
+}
+impl ManagedAsset {
+    /// Used to report the status of an Asset
+    pub fn report_status(_status: Result<(), ConfigError>) {}
+}
+
+/// Azure Device Registry Dataset that includes additional functionality
+/// to report status, translate data, and send data to the destination
+pub struct ManagedDataset<T: DataTransformer> {
+    /// Dataset Definition
+    pub dataset_definition: AssetDataset,
+    data_transformer: T,
+    reporter: Arc<Reporter>,
+}
+#[allow(dead_code)]
+impl<T> ManagedDataset<T>
+where
+    T: DataTransformer,
+{
+    pub(crate) fn new(dataset_definition: AssetDataset) -> Self {
+        // Create a new dataset
+        let forwarder = Forwarder::new(dataset_definition.clone());
+        let reporter = Arc::new(Reporter::new(dataset_definition.clone()));
+        let data_transformer = T::new(dataset_definition.clone(), forwarder, reporter.clone());
```

Comment:
Worth noting that this now has a transformer per dataset. I think I prefer that design decision, for a few reasons, but it's different than what we discussed before.